File: app/src/main/java/com/anglersparadise/ui/tank/FishTankView.java
// app/src/main/java/com/anglersparadise/ui/tank/FishTankView.java
package com.anglersparadise.ui.tank;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.view.View;

import com.anglersparadise.domain.model.Fish;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

public class FishTankView extends View {

    private List<Fish> fish = java.util.Collections.emptyList();
    private List<Fish> pendingFish = null;

    private final Map<Long, Float> xPos = new HashMap<>();
    private final Map<Long, Float> yPos = new HashMap<>();
    private final Map<Long, Float> speed = new HashMap<>();
    private final Random rng = new Random();

    public FishTankView(Context context) { this(context, null); }
    public FishTankView(Context context, AttributeSet attrs) { super(context, attrs); }

    public void setFish(List<Fish> list) {
        if (getWidth() == 0 || getHeight() == 0) {
            fish = list;
            pendingFish = list;
            invalidate();
            return;
        }
        applyFishList(list);
    }

    private void applyFishList(List<Fish> list) {
        fish = list;
        float w = Math.max(1f, getWidth());
        float h = Math.max(1f, getHeight());

        for (Fish f : list) {
            long id = f.getId();
            if (!xPos.containsKey(id)) {
                xPos.put(id, randFloat(-50f, w));
                yPos.put(id, randFloat(h * 0.2f, h * 0.8f));
                speed.put(id, randFloat(1.2f, 3.2f));
            }
        }
        // prune removed fish
        java.util.HashSet<Long> ids = new java.util.HashSet<>();
        for (Fish f : list) ids.add(f.getId());
        java.util.ArrayList<Long> toRemove = new java.util.ArrayList<>();
        for (Long id : xPos.keySet()) if (!ids.contains(id)) toRemove.add(id);
        for (Long id : toRemove) { xPos.remove(id); yPos.remove(id); speed.remove(id); }

        invalidate();
    }

    @Override protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        if (pendingFish != null) {
            applyFishList(pendingFish);
            pendingFish = null;
        }
    }

    @Override protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        if (fish.isEmpty()) return;

        float w = getWidth();
        float h = getHeight();

        for (Fish f : fish) {
            long id = f.getId();
            float vx = (speed.containsKey(id) ? speed.get(id) : 2f);
            float newX = (xPos.containsKey(id) ? xPos.get(id) : -50f) + vx;
            xPos.put(id, (newX > w + 50f) ? -50f : newX);
            float y = yPos.containsKey(id) ? yPos.get(id) : (h * 0.5f);

            // size -> sprite bounds
            float base = dp(48f);
            float incr = dp(8f) * (clamp(f.getSize(), 1, 5) - 1);
            int spriteW = Math.max(1, Math.round(base + incr));
            int spriteH = Math.max(1, Math.round(base * 0.7f + incr * 0.7f));

            android.graphics.Bitmap bmp =
                    SpriteProvider.get(getContext(), f.getSpecies(), spriteW, spriteH);

            float left = xPos.get(id);
            float top = y - spriteH / 2f;
            RectF dst = new RectF(left, top, left + spriteW, top + spriteH);
            canvas.drawBitmap(bmp, null, dst, null);
        }

        postInvalidateOnAnimation();
    }

    private float dp(float px) { return px * getResources().getDisplayMetrics().density; }
    private float randFloat(float a, float b) { return a + rng.nextFloat() * (b - a); }
    private int clamp(int v, int lo, int hi) { return Math.max(lo, Math.min(hi, v)); }
}

File: app/src/main/java/com/anglersparadise/ui/tank/FishTankView.kt
// app/src/main/java/com/anglersparadise/ui/tank/FishTankView.kt

package com.anglersparadise.ui.tank

import android.content.Context
import android.graphics.Canvas
import android.graphics.RectF
import android.util.AttributeSet
import android.view.View
import com.anglersparadise.domain.model.Fish
import kotlin.math.roundToInt

class FishTankView @JvmOverloads constructor(
    context: Context, attrs: AttributeSet? = null
) : View(context, attrs) {

    private var fish: List<Fish> = emptyList()
    private var pendingFish: List<Fish>? = null

    // Positions by fish id
    private val xPos = mutableMapOf<Long, Float>()
    private val yPos = mutableMapOf<Long, Float>()
    private val speed = mutableMapOf<Long, Float>()

    fun setFish(list: List<Fish>) {
        // If we don't have a measured size yet, defer initialization.
        if (width == 0 || height == 0) {
            fish = list
            pendingFish = list
            invalidate()
            return
        }
        applyFishList(list)
    }

    private fun applyFishList(list: List<Fish>) {
        fish = list
        val w = width.toFloat().coerceAtLeast(1f)
        val h = height.toFloat().coerceAtLeast(1f)

        // Initialize new fish positions safely
        for (f in list) {
            if (xPos[f.id] == null) {
                xPos[f.id] = (-50f..w).random()
                yPos[f.id] = (h * 0.2f .. h * 0.8f).random()
                speed[f.id] = (1.2f..3.2f).random()
            }
        }
        // Remove positions for fish no longer present
        val ids = list.map { it.id }.toSet()
        xPos.keys.toList().filter { it !in ids }.forEach { id ->
            xPos.remove(id); yPos.remove(id); speed.remove(id)
        }
        invalidate()
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        // If we had deferred data, initialize now that we have size.
        pendingFish?.let {
            applyFishList(it)
            pendingFish = null
        }
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (fish.isEmpty()) return

        val w = width.toFloat()
        val h = height.toFloat()

        for (f in fish) {
            val id = f.id
            val vx = (speed[id] ?: 2f)
            val newX = ((xPos[id] ?: -50f) + vx)
            // Wrap with a little margin
            xPos[id] = if (newX > w + 50f) -50f else newX
            val y = yPos[id] ?: (h * 0.5f)

            // Scale sprite by fish size (1..5)
            val base = dp(48f)
            val incr = dp(8f) * (f.size.coerceIn(1, 5) - 1)
            val spriteW = (base + incr).roundToInt().coerceAtLeast(1)
            val spriteH = (base * 0.7f + incr * 0.7f).roundToInt().coerceAtLeast(1)

            val bmp = SpriteProvider.get(context, f.species, spriteW, spriteH)

            val left = xPos[id] ?: 0f
            val top = y - spriteH / 2f
            val dst = RectF(left, top, left + spriteW, top + spriteH)
            canvas.drawBitmap(bmp, null, dst, null)
        }

        postInvalidateOnAnimation()
    }

    private fun dp(px: Float): Float = px * resources.displayMetrics.density
    private fun ClosedFloatingPointRange<Float>.random(): Float {
        return (start + Math.random() * (endInclusive - start)).toFloat()
    }
}
